Source/WebKit/Platform/IPC/Connection.h

```diff
@@ -392,9 +394,11 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
         DispatchMessagesFunction m_dispatchMessages;
         unsigned m_throttlingLevel { 0 };
     };
+    class SyncMessageState;
 
     static Lock s_connectionMapLock;
     Client* m_client { nullptr };
+    SyncMessageState* m_syncState { nullptr };
```

Comment:
I really don't like that this variable is protected by a lock, but there's no way to really know that except by code inspection of the callsites.

I understand that writes only happen from one thread, so we don't need to use the lock for those locations, but that's going to be somewhat hard for a future reader to know without careful analysis.

This is only going to get harder once we try to access data on the SyncMessageState (like the SerialFunctionDispatcher), since there will be even more callsites where it's hard to know what's safe to access, and how it can be guaranteed to be alive for.

I'm really a huge fan of clang's thread safety annotations (or DataMutex :)), since then the compiler can immediately tell you if you're adding a bug (for this subset of bugs). Obviously that means we have to lock on all threads, but my opinion is that this is a really good tradeoff.

If the performance cost of the extra locking really is a problem, then it seems possible to come up with a 'single writer, multiple readers' pattern (maybe a variant of DataMutex?) which makes it harder to get wrong. Can't determine safety statically, but we could at least have named accessors (like `accessFromWriterThread`) which force you to declare your intent.

Reply:
> I really don't like that this variable is protected by a lock, but there's no way to really know that except by code inspection of the callsites.

Yeah, it's not ideal, though it is an established pattern in WebKit code wrt "variable is mostly accessed from the maintop".

> If the performance cost of the extra locking really is a problem

It's mostly not a perf problem. It's a consistency problem. You look at the code and you see that clearly the variable is only ever modified from the main thread. Yet, you see all main thread read accesses lock the read. What's up with that.

You can contrast:
- No explanation of the locking: leaves the reader questioning.
- Excess locking: leaves the code inconsistent.

I added the static analysis support for the run-time properties.  How does the patch look now?

Reply:
I'm not sure I entirely understand the new static analysis, isn't the LockAssertion object always available (since it's a static const), so the static analysis can't catch failing cases? There's a lot of extra code added, and it's hard (for me) to understand what constraints it's able to enforce.

I guess I just feel like excess locking is less likely to leave the reader questioning. The variable would be locked for every access, and the compiler would fail if you ever wrote code that didn't do that. It's indeed unnecessary in some cases, but it's at least a very clearly enforced constraint, despite being a superset of the minimum 'safe' constraints.

I'm not yet a WebKit reviewer, so it might be best for @cdumez or someone to look at this time, since it'll require their sign off regardless.

Reply:
> I'm not sure I entirely understand the new static analysis, isn't the LockAssertion object always available (since it's a static const), so the static analysis can't catch failing cases? 

Static analysis of run-time properties cannot catch failing nor non-failing cases. That's why it's a run-time property, if if was not, it would be a compile-time property.  :) 

However, there is a test of the system proves that this catches non-trivial negative cases.  The catch is that it's a runtime error, e.g. an ASSERT on lock incorrectly not held.

It also catches trivial negative cases, e.g. missing assertions about the locks.

The system added here is the same as with asserting that code is run in particular thread:

```
struct {
   int n; 
   int m WTF_GUARDED_BY(m_workQueue);
} z;

void myFeatureThatIsRunOnWorkQueue() {
    assertIsCurrent(m_workQueue);
    z.m = z.n +1;
}
```

The compiler utterly just cannot know that the code is run in work queue.
What the compiler thinks:

- Programmer tells me that the code is run on the work queue.
- Accepting this as a fact, `z.m` can be assigned to since it can be only assigned when m_workQueue is current.

What the programmer thinks and tells the computer to do:
- I know the code is run only on the work queue.
- If during runtime we enter here, please ASSERT if we are not on the work queue.
- If I don't see the assertions, I'm confident I'm not lying to the compiler that the code is run on the work queue.

What the computer does during runtime:
- Check that the current work queue is the passed in work queue. ASSERT if not.

Similarly:
```
const LockAssertion& myTuesdaysLock() { if (isTuesday()) return LockAssertion::shared; return LockAssertion::unlocked; }
struct {
   int n WTF_ GUARDED_BY(myTuesdaysLock());
   int m;
} z;

void myFeatureThatIsRunOnTuesdays() {
   assertIsShared(myTuesdaysLock());
   z.m = z.n + 1;
}
```

Here the compiler thinks:
- Programmer tells me that the lock returned by `myTuesdaysLock()` is held as shared.
- Based on this, the assignment is correct to read the variable `z.n`, since that variable can be only read when `myTuesdaysLock()` is held as shared or exclusive.

What the programmer thinks and tells the computer to do:
- I know based on how I wrote the code that this code is only run on Tuesdays.
- Computer, please ASSERT if we are running the code on any other day than Tuesday.
- If I don't see the assertions, I'm confident I'm not lying to the compiler that the code is run on Tuesdays only.

What the computer does:
- Check that myTuesdaysLock() returns shared. ASSERT if not.


> There's a lot of extra code added, and it's hard (for me) to understand what constraints it's able to enforce.

It's able to enforce everything, but it's a bit cumbersome since it does not understand the underlying locks, rather it must rely on the explicit assertions (similar to the is current case). It is also vulnerable to the incorrectly specified lock conditions (like everything?).